src/Lykke.Service.ExchangeDataStore.AzureRepositories/ExchangeInstruments/ExchangeInstrumentsRepository.cs

```diff
@@ -0,0 +1,73 @@
+﻿using AzureStorage;
+using Common.Log;
+using Lykke.Service.ExchangeDataStore.Core.Domain.Exchange;
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Lykke.Service.ExchangeDataStore.AzureRepositories.ExchangeInstruments
+{
+    // ReSharper disable once ClassNeverInstantiated.Global - Autofac injected
+    public class ExchangeInstrumentsRepository : IExchangeInstrumentsRepository
+    {
+        private readonly ILog _log;
+        private static readonly string _className = nameof(ExchangeInstrumentsRepository);
+        private readonly INoSQLTableStorage<ExchangeInstrumentEntity> _tableStorage;
+        private readonly ConcurrentDictionary<string, IList<string>> _knownExchangeInstruments; //local cache, dont check in Azure multiple times for exchanges/instruments we already have data for
+        
+        public ExchangeInstrumentsRepository(INoSQLTableStorage<ExchangeInstrumentEntity> tableStorage, ILog log)
+        {
+            _tableStorage = tableStorage;
+            _log = log;
+            _knownExchangeInstruments = new ConcurrentDictionary<string, IList<string>>();
+        }
+
+        public async Task SaveIfNotExists(string exchangeName, string instrument)
+        {
+            if (!_knownExchangeInstruments.ContainsKey(exchangeName) || (_knownExchangeInstruments.TryGetValue(exchangeName, out var instruments) && !instruments.Contains(instrument)) )
+            {
+                var added = await _tableStorage.CreateIfNotExistsAsync(new ExchangeInstrumentEntity(exchangeName, instrument));
+                if (added)
+                {
+                    await _log.WriteInfoAsync(_className, nameof(SaveIfNotExists), $"{exchangeName} & {instrument} added to ExchangeInstruments table.");
+                    _knownExchangeInstruments.AddOrUpdate(exchangeName, new List<string> { instrument }, (key, oldValue) =>
```

Comment:
Well, it would have added it twice, you are right. However, it would never get to that point, because of the IF check at the top of the method:
 if ( ........ && !instruments.Contains(instrument)) ). The top IF check ensures that we add exchange/instrument to the local cache only if the exchange (key) is not in the dictionary or the instrument is not in the list of instruments for that exchange. Nevertheless, I followed your suggestion and added the extra check to only add the instrument to the value if its not there for consistency and to avoid confision:

if(!oldValue.Contains(instrument)) oldValue.Add(instrument);
return oldValue;

There was another problem I found: _knownExchangeInstruments.AddOrUpdate code was inside the if(add) section, which would have prevent it being executed for exchanges/instruments that are already in the DB ( if(add) would be false for them and they would never get added to the cache, so we would have queried the DB each time ). I have taken it outside now and tested it - cache works as expected - no duplicates and no unnecessary queries to the DB.